packages/next/src/server/use-cache/use-cache-wrapper.ts

```diff
@@ -791,3 +795,33 @@ export function cache(
   }[name]
   return cachedFn
 }
+
+function createLazyPromise<TResult>(
+  fn: () => Promise<TResult>
+): Promise<TResult> {
+  let invoked = false
+  let resolveFn: (value: TResult) => void
+  let rejectFn: (reason?: unknown) => void
+
+  const promise = new Promise<TResult>((resolve, reject) => {
+    resolveFn = resolve
+    rejectFn = reject
+  })
+
+  return new Proxy(promise, {
+    get(target, prop, receiver) {
+      if (prop === 'then') {
+        return (...args: unknown[]) => {
+          if (!invoked) {
+            invoked = true
+            fn().then(resolveFn, rejectFn)
+          }
+
+          return Reflect.apply(target.then, target, args)
+        }
+      }
+
+      return Reflect.get(target, prop, receiver)
+    },
+  })
+}
```

Comment:
This is neat but maybe overkill. can just override the `then` property on a promise instance or create a thenable type since React treats all thenables as castable to promises.

```
function createLazyResult(fn) {
  let pendingResult;
  return {
    then(onFullfilled, onErrored) {
      if (!pendingResult) {
        pendingResult = fn()
      }
      pendingResult.then(onFullfilled, onErrored)
    }
  }
}
```

Reply:
Hm, I thought about only patching `then`, which would be enough for React (at least today). But I also imagined this being extracted into a shared helper soon to possibly be used for server actions as well. And I wanted to avoid the surprising runtime error if someone (in the Next.js code base) ever calls `.catch()` on the `resultPromise` here, or on the `actionResult`. But if we type them as `PromiseLike` instead of `Promise`, I think this should be fine, so why not.

Reply:
Yeah I'd do that personally. you can always turn it into a promise with the .then if you need access to a finally and if you want to handle the rejection there's always the second arg to .then so as long as the type system hints to you what you need to do this seems preferable to avoid the proxy